### ISSUE TITLE
[FEAT] Different clustering algorithms besides Connexted Components

### ISSUE THREAD

**Issue by NickCrews** [moj-analytical-services/splink#1824]
### Is your proposal related to a problem?
I was reading https://dbs.uni-leipzig.de/files/research/publications/2021-3/pdf/Paper_MSCD_AP_BTW21_Dresden.pdf and I noticed in figure 4 that Connected Components ranks poorly on the f1 score when compared to other clustering algorithms, on all three example datasets. I worry that I and other users of splink are getting great comparison performance, where a lot of development effort is going into splink, but we are tossing out a lot of accuracy in the final mile with the clustering algorithm.

### Describe the solution you'd like
Not exactly sure. Probably one of the other algorithms described there? Needs more research than that.

Probably the connected components should be implemented in SQL as it is now (because it needs to be scalable), but then within each cluster (hopefully less than 1k records??) we can run a clustering algorithm implemented in whatever language we want, eg sklearn.

### Describe alternatives you've considered
Need to think of this mroe.

**Comment by RossKen** [comment]
Hey @NickCrews - thanks for raising this issue.

This brings up some interesting questions about how people use Splink and what problems they are solving. We have discussed this internally in the past and concluded that the explainability of connected components is one of its key strengths. Given we, primarily, use Splink in a government setting, that transparency is particularly important to us. 

We are open to the idea of using different clustering methodologies, particularly if there is a performance boost, but up to this point there have always been other developments which we have seen as higher priority. 

I think this taps into a wider point on how we measure performance more generally. We are aware that we have various metrics and methods for assessing performance at the pairwise comparison level, but not very much at the cluster level. This is something we aim to tackle this year, so how we create clusters may well get looked at in more depth at that stage.
